<div style="background:linear-gradient(145deg,#111827 0%,#1F2937 55%,#172554 100%); padding:25px; border-radius:20px; border:1px solid #475569; box-shadow:0 10px 30px rgba(0,0,0,0.3);">

<div style="font-size:12px; color:#60A5FA; letter-spacing:2px; margin-bottom:8px;">
◈ PRACTICAL LAB / MODEL VALIDATION
</div>

<h1 style="color:#F8FAFC; font-size:30px; margin:0; line-height:1.3;">
🧪 Hands-On Lab
<br>
<span style="color:#60A5FA;">Cross-Validating a Model</span>
</h1>

<div style="height:3px; width:75px; background:#38BDF8; margin-top:16px; border-radius:5px;"></div>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0; margin-top:16px;">
In this hands-on lab, we will apply cross-validation to a classification model
and evaluate how consistently it performs across different validation folds.
We will then compare the cross-validated performance with the single-split
evaluation from the previous work.
</p>

<div style="display:flex; gap:10px; margin-top:20px;">

<div style="flex:1; background:#0F172A; padding:13px; border-radius:12px; border:1px solid #334155;">
<div style="color:#64748B; font-size:10px;">MODEL</div>
<div style="color:#E2E8F0; font-size:13px; margin-top:5px;">🌳 Decision Tree</div>
</div>

<div style="flex:1; background:#0F172A; padding:13px; border-radius:12px; border:1px solid #334155;">
<div style="color:#64748B; font-size:10px;">VALIDATION</div>
<div style="color:#E2E8F0; font-size:13px; margin-top:5px;">5-Fold CV</div>
</div>

<div style="flex:1; background:#0F172A; padding:13px; border-radius:12px; border:1px solid #334155;">
<div style="color:#64748B; font-size:10px;">METRIC</div>
<div style="color:#E2E8F0; font-size:13px; margin-top:5px;">F1 Score</div>
</div>

</div>

</div>

<div style="background:#0F172A; padding:20px; border-radius:15px; border-left:5px solid #60A5FA;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:2px;">
HANDS-ON / STEP 01
</div>

<h2 style="color:#F8FAFC; margin:8px 0 10px 0;">
🔄 Evaluate the Model with 5-Fold Cross-Validation
</h2>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0;">

We will take the <b style="color:#93C5FD;">Decision Tree Classifier</b> and
evaluate it using <b style="color:#93C5FD;">5-fold cross-validation</b>.

Instead of relying on one validation split, the training data will be divided
into five folds. The model will be trained and validated five times, with a
different fold used for validation each time.

The <b style="color:#93C5FD;">F1-score</b> will be used to measure the model's
performance on each fold.

</p>

</div>

In [27]:
import pandas as pd

df = pd.read_csv("bank-additional-full.csv", sep=";")

print("Dataset loaded successfully!")
print(f"Shape: {df.shape}")

Dataset loaded successfully!
Shape: (41188, 21)


In [28]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

data = df.copy()

data = data.drop(columns=["duration"])

categorical_cols = data.select_dtypes(include=["object"]).columns

label_encoders = {}

for col in categorical_cols:
    le = LabelEncoder()
    data[col] = le.fit_transform(data[col])
    label_encoders[col] = le

X = data.drop(columns=["y"])
y = data["y"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("X_train shape:", X_train.shape)
print("y_train shape:", y_train.shape)
print("X_test shape:", X_test.shape)
print("y_test shape:", y_test.shape)

X_train shape: (32950, 19)
y_train shape: (32950,)
X_test shape: (8238, 19)
y_test shape: (8238,)


C:\Users\lenovo\AppData\Local\Temp\ipykernel_18356\2058256427.py:8: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = data.select_dtypes(include=["object"]).columns


In [29]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

model = DecisionTreeClassifier(random_state=42)

cv_scores = cross_val_score(
    model,
    X_train,
    y_train,
    cv=5,
    scoring="f1"
)

print("5-Fold F1 Scores:")
print(cv_scores)

5-Fold F1 Scores:
[0.30729834 0.3012987  0.32611465 0.32470893 0.3073955 ]


<div style="background:linear-gradient(135deg,#0F172A,#172554); padding:20px; border-radius:15px; border:1px solid #334155;">

<h3 style="color:#93C5FD; margin-top:0;">
📊 What Did We Learn?
</h3>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8;">

The model produced slightly different F1-scores across the five folds.
This is expected because each fold contains a different subset of the
training data.

<br><br>

The scores range from approximately <b style="color:#60A5FA;">0.3013</b>
to <b style="color:#60A5FA;">0.3261</b>, showing that the model's performance
does not change dramatically between folds.

<br><br>

This gives us a more reliable view of the model's performance than relying
on only one train/validation split.

</p>

<div style="margin-top:15px; padding:12px; background:#111827; border-radius:10px; border:1px solid #334155; color:#86EFAC; font-size:13px;">
● 5 folds completed successfully
</div>

</div>

<div style="background:#0F172A; padding:20px; border-radius:15px; border-left:5px solid #60A5FA;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:2px;">
HANDS-ON / STEP 02
</div>

<h2 style="color:#F8FAFC; margin:8px 0 10px 0;">
📊 Calculate Mean & Standard Deviation
</h2>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0;">

After obtaining an F1-score from each fold, we need to summarize the
cross-validation results.

The <b style="color:#93C5FD;">mean</b> represents the average performance
of the model across all five folds, while the
<b style="color:#93C5FD;">standard deviation</b> measures how much the
performance varies between folds.

</p>

</div>

In [30]:

import numpy as np

mean_f1 = np.mean(cv_scores)
std_f1 = np.std(cv_scores)

print(f"Mean F1-Score: {mean_f1:.4f}")
print(f"Standard Deviation: {std_f1:.4f}")

Mean F1-Score: 0.3134
Standard Deviation: 0.0101


In [31]:
from IPython.display import display, HTML

display(HTML(f"""
<div style="
    background:linear-gradient(145deg,#111827,#172554);
    padding:22px;
    border-radius:16px;
    border:1px solid #334155;
    box-shadow:0 8px 22px rgba(0,0,0,0.25);
">

<div style="
    color:#64748B;
    font-size:11px;
    letter-spacing:2px;
    margin-bottom:14px;
">
STEP 02 / CV PERFORMANCE SUMMARY
</div>

<div style="display:flex; gap:12px;">

<div style="
    flex:1;
    background:#0F172A;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">MEAN F1-SCORE</div>
    <div style="color:#60A5FA; font-size:26px; font-weight:bold; margin-top:6px;">
        {mean_f1:.4f}
    </div>
    <div style="color:#64748B; font-size:11px; margin-top:4px;">
        Average across 5 folds
    </div>
</div>

<div style="
    flex:1;
    background:#0F172A;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">STANDARD DEVIATION</div>
    <div style="color:#93C5FD; font-size:26px; font-weight:bold; margin-top:6px;">
        {std_f1:.4f}
    </div>
    <div style="color:#64748B; font-size:11px; margin-top:4px;">
        Variation between folds
    </div>
</div>

</div>

<div style="
    margin-top:15px;
    padding:12px;
    background:#172554;
    border-radius:10px;
    border:1px solid #334155;
    color:#CBD5E1;
    font-size:13px;
">
● The low standard deviation indicates relatively consistent performance
across the five validation folds.
</div>

</div>
"""))

<div style="background:#0F172A; padding:20px; border-radius:15px; border:1px solid #334155;">

<h3 style="color:#93C5FD; margin-top:0;">
🔎 Interpreting the Results
</h3>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8;">

The Decision Tree achieved a mean F1-score of
<b style="color:#60A5FA;">0.3134</b> across the five folds.

The standard deviation is only
<b style="color:#60A5FA;">0.0101</b>, which indicates that the F1-score
did not vary greatly from one fold to another.

<br><br>

Therefore, the model shows relatively
<b style="color:#86EFAC;">stable performance</b> across different subsets
of the training data.

However, the relatively low F1-score means that the model's overall
classification performance is not particularly strong, so stability
does not necessarily mean high predictive performance.

</p>

</div>

<div style="background:#0F172A; padding:20px; border-radius:15px; border-left:5px solid #60A5FA;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:2px;">
HANDS-ON / STEP 03
</div>

<h2 style="color:#F8FAFC; margin:8px 0 10px 0;">
🔍 Cross-Validation vs. Single Split
</h2>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0;">

In this step, we compare the model's performance from a
<b style="color:#93C5FD;">single train/test split</b> with its average
performance from <b style="color:#93C5FD;">5-fold cross-validation</b>.

This helps us understand whether the original single-split result was
representative of the model's general performance or was influenced by
one particular data split.

</p>

</div>

In [32]:
from sklearn.metrics import f1_score

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

single_split_f1 = f1_score(y_test, y_pred)

print(f"Single-Split F1-Score: {single_split_f1:.4f}")

Single-Split F1-Score: 0.3311


In [33]:
display(HTML(f"""
<div style="
    background:linear-gradient(145deg,#111827,#172554);
    padding:22px;
    border-radius:16px;
    border:1px solid #334155;
    box-shadow:0 8px 22px rgba(0,0,0,0.25);
">

<div style="
    color:#64748B;
    font-size:11px;
    letter-spacing:2px;
">
SINGLE-SPLIT EVALUATION
</div>

<div style="
    color:#60A5FA;
    font-size:30px;
    font-weight:bold;
    margin-top:8px;
">
{single_split_f1:.4f}
</div>

<div style="
    color:#94A3B8;
    font-size:12px;
    margin-top:4px;
">
F1-score on the held-out test set
</div>

</div>
"""))

In [34]:
display(HTML(f"""
<div style="
    background:#0F172A;
    padding:20px;
    border-radius:16px;
    border:1px solid #334155;
">

<div style="color:#64748B; font-size:11px; letter-spacing:2px;">
PERFORMANCE COMPARISON
</div>

<div style="display:flex; gap:12px; margin-top:15px;">

<div style="
    flex:1;
    background:#111827;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        SINGLE SPLIT
    </div>
    <div style="color:#60A5FA; font-size:24px; font-weight:bold; margin-top:5px;">
        {single_split_f1:.4f}
    </div>
</div>

<div style="
    flex:1;
    background:#111827;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        5-FOLD CV MEAN
    </div>
    <div style="color:#93C5FD; font-size:24px; font-weight:bold; margin-top:5px;">
        {mean_f1:.4f}
    </div>
</div>

<div style="
    flex:1;
    background:#111827;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        DIFFERENCE
    </div>
    <div style="color:#86EFAC; font-size:24px; font-weight:bold; margin-top:5px;">
        {abs(single_split_f1 - mean_f1):.4f}
    </div>
</div>

</div>

</div>
"""))

<div style="background:#0F172A; padding:20px; border-radius:15px; border:1px solid #334155;">

<h3 style="color:#93C5FD; margin-top:0;">
🔎 Single Split vs. Cross-Validation
</h3>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8;">

The single train/test split produced an F1-score of
<b style="color:#60A5FA;">0.3311</b>, while the 5-fold cross-validation
produced a mean F1-score of <b style="color:#60A5FA;">0.3134</b>.

<br><br>

The difference between the two results is
<b style="color:#60A5FA;">0.0177</b>. The single-split score is slightly
higher, which suggests that the particular test split used in the original
evaluation was somewhat more favorable to the model than the average
validation fold.

<br><br>

Cross-validation gives us a more reliable estimate because it evaluates the
model across multiple different subsets of the training data instead of
depending on one particular split.

</p>

<div style="margin-top:15px; padding:12px; background:#172554; border-radius:10px; border:1px solid #334155; color:#93C5FD; font-size:13px;">
💡 The cross-validation mean provides a broader view of expected model performance.
</div>

</div>

<div style="background:#0F172A; padding:20px; border-radius:15px; border-left:5px solid #60A5FA;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:2px;">
HANDS-ON / STEP 04
</div>

<h2 style="color:#F8FAFC; margin:8px 0 10px 0;">
⚖️ Confirm Stratified Folds
</h2>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8; margin-bottom:0;">

Because our task is a <b style="color:#93C5FD;">classification problem</b>,
we need to make sure that each validation fold maintains a similar
distribution of the target classes.

Scikit-learn automatically uses <b style="color:#93C5FD;">StratifiedKFold</b>
when <code>cross_val_score</code> receives a classifier and an integer value
such as <code>cv=5</code>.

We will explicitly create a <b style="color:#93C5FD;">StratifiedKFold</b>
object and use it for cross-validation to confirm this behavior.

</p>

</div>

In [35]:
from sklearn.model_selection import StratifiedKFold

stratified_cv = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

print(stratified_cv)

StratifiedKFold(n_splits=5, random_state=42, shuffle=True)


In [36]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

model = DecisionTreeClassifier(random_state=42)

cv_scores = cross_val_score(
    model,
    X_train,
    y_train,
    cv=5,
    scoring="f1"
)

cv_scores

array([0.30729834, 0.3012987 , 0.32611465, 0.32470893, 0.3073955 ])

In [37]:
from IPython.display import display, HTML

display(HTML(f"""
<div style="
    background:#0F172A;
    padding:22px;
    border-radius:18px;
    border:1px solid #334155;
    box-shadow:0 8px 25px rgba(0,0,0,0.25);
">

<div style="
    color:#64748B;
    font-size:11px;
    letter-spacing:2px;
">
STRATIFIED CROSS-VALIDATION
</div>


<div style="
    color:#F8FAFC;
    font-size:22px;
    font-weight:bold;
    margin-top:6px;
">
⚖️ 5-Fold F1 Performance
</div>


<div style="
    display:flex;
    gap:10px;
    margin-top:20px;
">


<div style="
    flex:1;
    background:#111827;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        FOLD 01
    </div>
    <div style="color:#60A5FA; font-size:21px; font-weight:bold; margin-top:5px;">
        {stratified_scores[0]:.4f}
    </div>
</div>


<div style="
    flex:1;
    background:#111827;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        FOLD 02
    </div>
    <div style="color:#60A5FA; font-size:21px; font-weight:bold; margin-top:5px;">
        {stratified_scores[1]:.4f}
    </div>
</div>


<div style="
    flex:1;
    background:#111827;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        FOLD 03
    </div>
    <div style="color:#60A5FA; font-size:21px; font-weight:bold; margin-top:5px;">
        {stratified_scores[2]:.4f}
    </div>
</div>


<div style="
    flex:1;
    background:#111827;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        FOLD 04
    </div>
    <div style="color:#60A5FA; font-size:21px; font-weight:bold; margin-top:5px;">
        {stratified_scores[3]:.4f}
    </div>
</div>


<div style="
    flex:1;
    background:#111827;
    padding:14px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#64748B; font-size:10px;">
        FOLD 05
    </div>
    <div style="color:#60A5FA; font-size:21px; font-weight:bold; margin-top:5px;">
        {stratified_scores[4]:.4f}
    </div>
</div>


</div>


<div style="
    display:flex;
    gap:12px;
    margin-top:14px;
">


<div style="
    flex:1;
    background:#172554;
    padding:16px;
    border-radius:12px;
    border:1px solid #3B82F6;
">
    <div style="color:#93C5FD; font-size:10px;">
        MEAN F1-SCORE
    </div>
    <div style="color:#F8FAFC; font-size:26px; font-weight:bold; margin-top:5px;">
        {mean_f1:.4f}
    </div>
</div>


<div style="
    flex:1;
    background:#111827;
    padding:16px;
    border-radius:12px;
    border:1px solid #334155;
">
    <div style="color:#94A3B8; font-size:10px;">
        STANDARD DEVIATION
    </div>
    <div style="color:#F8FAFC; font-size:26px; font-weight:bold; margin-top:5px;">
        {std_f1:.4f}
    </div>
</div>


</div>


<div style="
    margin-top:16px;
    padding-top:13px;
    border-top:1px solid #334155;
    color:#94A3B8;
    font-size:12px;
">
    ⚖️ Stratified validation preserves class proportions across folds.
</div>


</div>
"""))

<div style="background:linear-gradient(145deg,#111827,#172554); padding:22px; border-radius:17px; border:1px solid #334155;">

<div style="color:#60A5FA; font-size:12px; letter-spacing:2px;">
STRATIFIED CROSS-VALIDATION / RESULTS
</div>

<h2 style="color:#F8FAFC; margin:10px 0;">
⚖️ What Do the Results Tell Us?
</h2>

<p style="color:#CBD5E1; font-size:15px; line-height:1.8;">

The Decision Tree achieved a mean F1-score of
<b style="color:#60A5FA;">0.3205</b> using 5-fold Stratified Cross-Validation.

The standard deviation was
<b style="color:#60A5FA;">0.0114</b>, indicating that the model's performance
remained relatively consistent across the five stratified folds.

<br><br>

The folds preserved a similar proportion of the target classes, which is
especially important because the Bank Marketing dataset is an
<b style="color:#93C5FD;">imbalanced classification problem</b>.

<br><br>

Compared with the previous cross-validation mean of
<b style="color:#60A5FA;">0.3134</b>, the stratified result is very similar.
This confirms that the classification performance is reasonably consistent
when the class distribution is preserved across folds.

</p>

<div style="margin-top:16px; padding:13px; background:#0F172A; border-radius:10px; border:1px solid #334155; color:#86EFAC; font-size:13px;">
● Stratified 5-Fold Validation Completed Successfully
</div>

</div>

In [38]:
StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

StratifiedKFold(n_splits=5, random_state=42, shuffle=True)